README.md
# MaybeXml

MaybeXml is a library to scan and evaluate [XML][xml]-like data into tokens. In
effect, the library provides a non-validating lexer. The interface is similar to
many XML pull parsers.

* [Latest API Documentation][api_docs]

## Purpose

The purpose of the library is to provide a way to read XML documents including
office suite documents, RSS/Atom feeds, config files, SVG, and web service
messages.

## Usage

The library user creates a `Lexer` from a `&str`.

Then, the library user can call `Lexer::tokenize()` to try to get the next
`Token`. If successful, repeat calling `tokenize` and process the available
tokens.

Alternatively, the user can turn the `Lexer` into an iterator via
`Lexer::iter()` or `IntoIterator::into_iter()`.

## Examples

### Using `Lexer::tokenize()`

```rust
use maybe_xml::{Lexer, token::{Characters, EndTag, StartTag, Ty}};

let input = "<id>123</id>";

let lexer = Lexer::from_str(input);
let mut pos = 0;

let token = lexer.tokenize(&mut pos);
assert_eq!(Some(Ty::StartTag(StartTag::from_str("<id>"))), token.map(|t| t.ty()));
assert_eq!(4, pos);

let token = lexer.tokenize(&mut pos);
assert_eq!(Some(Ty::Characters(Characters::from_str("123"))), token.map(|t| t.ty()));
assert_eq!(7, pos);

let token = lexer.tokenize(&mut pos);
assert_eq!(Some(Ty::EndTag(EndTag::from_str("</id>"))), token.map(|t| t.ty()));
assert_eq!(12, pos);

let token = lexer.tokenize(&mut pos);
assert_eq!(None, token);

// Verify that `pos` is equal to `input.len()` to ensure all data was
// processed.
```

### Using `Iterator` functionality

```rust
use maybe_xml::{Lexer, token::{Characters, EndTag, StartTag, Ty}};

let input = "<id>Example</id>";

let lexer = Lexer::from_str(input);

let mut iter = lexer.into_iter().map(|token| token.ty());

let token_type = iter.next();
assert_eq!(token_type, Some(Ty::StartTag(StartTag::from_str("<id>"))));
match token_type {
    Some(Ty::StartTag(start_tag)) => {
        assert_eq!(start_tag.name().as_str(), "id");
    }
    _ => panic!("unexpected token"),
}
assert_eq!(Some(Ty::Characters(Characters::from_str("Example"))), iter.next());
assert_eq!(Some(Ty::EndTag(EndTag::from_str("</id>"))), iter.next());
assert_eq!(None, iter.next());
```

## Installation

```sh
cargo add maybe_xml
```

By default, the `std` feature is enabled.

### Alloc only

If the host environment has an allocator but does not have access to the Rust
`std` library:

```sh
cargo add --no-default-features --features alloc maybe_xml
```

### No allocator / core only

If the host environment does not have an allocator:

```sh
cargo add --no-default-features maybe_xml
```

## License

Licensed under either of [Apache License, Version 2.0][LICENSE_APACHE] or [MIT
License][LICENSE_MIT] at your option.

### Contributions

Unless you explicitly state otherwise, any contribution intentionally submitted
for inclusion in the work by you, as defined in the Apache-2.0 license, shall be
dual licensed as above, without any additional terms or conditions.

[LICENSE_APACHE]: LICENSE-APACHE
[LICENSE_MIT]: LICENSE-MIT
[xml]: https://www.w3.org/TR/2006/REC-xml11-20060816/
[api_docs]: https://docs.rs/maybe_xml/

CHANGELOG.md
# CHANGELOG

## [0.10.1] - 2023-12-01

### Fixed

- Fix incorrect parsing CDATA issue. The `pos` was not used when scanning for
  the CDATA.
  Thanks [@Mingun](https://github.com/Mingun).
  See [Issue #8](https://github.com/bluk/maybe_xml/issues/8)

## [0.10.0] - 2023-11-21

### Removed

- Removed deprecated functionality.

## [0.9.0] - 2023-11-14

### Updated

* **Breaking change**: `Lexer::into_inner()` returns a `&str`.
* **Breaking change**: `Lexer::tokenize()` restores the panic behavior in
  previous versions. If the position argument is greater than the input's byte length,
  then a panic occurs. The position argument is also immediately checked to be
  at a UTF-8 character boundary instead of waiting for a panic in a subsequent
  function call.
* **Breaking change**: Spaces around properties like attribute names, attribute
  values, and processing instructions are trimmed in their representations.
* **Breaking change**: Minimum supported Rust version is now `1.71.0`.
* Renamed `Lexer` into `Reader`. `Lexer` is still exposed as a hidden alias but will
  be removed in a future version.
* `Token::ty()` is made `const`.

### Removed

* Remove `From` and `from_str` implementations from token
  and property types. Due to possible future implementation changes and
  undocumented invariants, the ability to directly construct a type is removed.
* Remove `as_str()`, `as_bytes()`, `as_str_unchecked()`,
  and `into_inner()` implementations for `Attributes` and `Attribute`. Both are opaque
  types where the internal data may change.

### Deprecated

- Deprecated `from_slice_unchecked()` on `Lexer`. Use `core::str::from_utf8_unchecked()`
  on the byte slice and then pass into `Reader::from_str()` if the byte slice is
  guaranteed to be a valid UTF-8 string.

### Added

- Most functions are `const` on the token and property types. Refactored most
  token and property type function implementations to be `const` compatible.
- Add `Reader::parse()` as a `const` alternative function to
  `Reader::tokenize()`. The function does not mutate the position argument so the
  position must be re-calculated like `pos += token.len()` in the calling code.
- Add `Attributes.parse()` as a `const` function to parse the attributes like
  `Reader::parse()`.
- Add `Display` implementations to `Token` and token types.
- Derive `PartialOrd`, `Ord`, and `Hash` for `Token` and `Ty`.
- Add `len()` method to token types.

## [0.8.0] - 2023-11-11

### Added

- Add `Token::as_str()` and `as_str()` to other token and property types.
  `as_str()` replaces `to_str()` or `as_str_unchecked()` as the preferred method
  to get the underlying value.
- Add `Token::from_str()` and `from_str()` to other token and property types.
  The method should be used for constructing values in tests.

### Deprecated

- Deprecated `as_str_unchecked()`, `to_str()` and `into_inner()` on `Token`.
- Deprecated `as_str_unchecked()`, `to_str()` and `into_inner()` on token types.
- Deprecated `as_str_unchecked()` and `to_str()` on property types.

### Removed

- Remove `From<&[u8]>` and `AsRef<[u8]>` implementations from token and property
  types.

### Changed

- Changed the internal implementation of tokens to use `&str` instead of
  `&[u8]`. Micro benchmark performance should be similar to the previous
  implementation, but when using `as_str()` instead of `to_str()`, the calling
  code does not have to handle a `Result` type which should increase overall
  performance.

## [0.7.2] - 2023-11-01

### Changed

- Make `Iterator` from `Lexer::iter()` into an explicit type named `Iter`.

### Added

- Add `Lexer::into_inner()` to get the underlying bytes passed to the `Lexer`.
- Derive `PartialEq`, `Eq`, and `Hash` for `Lexer` and iterators.

## [0.7.1] - 2023-10-31

### Changed

- Modify `IntoIter` to be public.

## [0.7.0] - 2023-10-30

### Changed

- Add `Lexer::from_str` as the safe instantiation method.
- Refactor `Lexer::from_slice` to unsafe `Lexer::from_slice_unchecked`. It is
  assumed the byte slice is UTF-8 characters, and if they are not a valid slice
  of UTF-8 bytes, then the behavior is undefined.
- Dynamically determine `Token` type instead of storing the `Ty`. This reduces the
  `Token` size to 2 words (16 octets on a 64-bit machine).

### Removed

- Remove `offset` field from `Token`. If required, the offset should be
  calculated when tokens are consumed.
- Remove `scan` and `Scanner` from the public API.
- Remove `const` from `Token::ty()` method for possible future optimizations.
- Remove public `new` methods from `Token` and related types.

### Added

- Add unsafe `as_str_unchecked` to various types in cases where the bytes are
  guaranteed to be valid UTF-8.
- Add `as_bytes`, `to_str`, `as_str_unchecked`, and `into_inner` to `Token`.

## [0.6.0] - 2023-10-17

### Changed

- Change the minimum Rust supported version to `1.56.0` by moving benchmark code
  to own crate.

### Removed

- Removed deprecated functionality.

## [0.5.0] - 2023-09-28

### Added

- Add `Lexer` type which is the replacement for all evaluators. It borrows the
  input (a slice of bytes) and tokenizes with a borrowed `Token`. It has the
  benefits of both the existing evaluators (namely borrowing byte slices from the
  input so no allocations are necessary and allowing an `Iterator` API).
- Add `scan` function which scans for a token at the beginning of a slice.

### Deprecated

- Add deprecations to owned token types and all evaluators. The code will be
  removed in a future version.

### Fixed

- Fixed wrong lifetimes in the borrowed token types. The lifetimes should have
  been tied to the borrowed input instead of the token's type.

### Changed

- Derive most trait implementations for token and property types.
- Change the minimum Rust supported version to `1.60.0` by changing newer
  syntax code like `let...else` to older supported syntax.
- Add `#[must_use]` and `const` to functions.

### Removed


src/lib.rs
// Copyright 2022 Bryant Luk
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.

//! `MaybeXml` is a library to scan and evaluate [XML][xml]-like data into tokens. In
//! effect, the library provides a non-validating lexer. The interface is similar to many
//! XML pull parsers.
//!
//! # Usage
//!
//! The library user creates a [`Lexer`] from a slice of bytes. The slice of
//! bytes is usually from a buffer managed by the library user. For instance, it
//! could be a buffer of data that is being read over a network socket, a memory
//! mapped file, or just a [`Vec`] of bytes.
//!
//! Then, the library user can call [`Lexer::tokenize()`][`Lexer::tokenize()`]
//! to try to get the next [`Token`][crate::token::Token]. If successful,
//! repeat calling `tokenize` and process the available tokens.
//!
//! Alternatively, the user can turn the `Lexer` into an iterator via
//! [`Lexer::iter()`][Lexer::iter()] or [`IntoIterator::into_iter()`].
//!
//! # Examples
//!
//! ## Using [`Iterator`] functionality
//!
//! ```
//! use maybe_xml::{Lexer, token::{Characters, EndTag, StartTag, Token, Ty}};
//!
//! let input = b"<id>Example</id>";
//!
//! let lexer = Lexer::from_slice(input);
//!
//! let mut iter = lexer.into_iter().map(|token| token.ty());
//!
//! let token_type = iter.next();
//! assert_eq!(token_type, Some(Ty::StartTag(StartTag::from("<id>"))));
//! match token_type {
//!     Some(Ty::StartTag(start_tag)) => {
//!         assert_eq!(start_tag.name().to_str()?, "id");
//!     }
//!     _ => panic!("unexpected token"),
//! }
//! assert_eq!(iter.next(), Some(Ty::Characters(Characters::from("Example"))));
//! assert_eq!(iter.next(), Some(Ty::EndTag(EndTag::from("</id>"))));
//! assert_eq!(iter.next(), None);
//! # Ok::<(), core::str::Utf8Error>(())
//! ```
//!
//! ## Using [`Lexer::tokenize()`][Lexer::tokenize()] directly
//!
//! ```
//! use maybe_xml::{Lexer, token::{Characters, EndTag, StartTag, Ty}};
//!
//! let mut buf = Vec::new();
//! // Note the missing closing tag character `>` in the end tag.
//! buf.extend(b"<id>123</id");
//!
//! let lexer = Lexer::from_slice(&buf);
//! let mut pos = 0;
//!
//! let token = lexer.tokenize(&mut pos).unwrap();
//! assert_eq!(0, token.offset());
//! assert_eq!(Ty::StartTag(StartTag::from("<id>".as_bytes())), token.ty());
//!
//! let token = lexer.tokenize(&mut pos).unwrap();
//! assert_eq!(4, token.offset());
//! assert_eq!(Ty::Characters(Characters::from("123".as_bytes())), token.ty());
//!
//! let token = lexer.tokenize(&mut pos);
//! // The last token is incomplete because it is missing the `>`
//! assert_eq!(None, token);
//!
//! // The position was updated as tokenize() was called
//! assert_eq!(7, pos);
//!
//! // Remove the tokenized data and reset the position
//! buf.drain(..pos);
//! pos = 0;
//!
//! // Verify that the buffer is empty. If it is not empty, then there is data
//! // which could not be identified as a complete token. This usually indicates
//! // an error has occurred. If there is more data (say coming from a network
//! // socket), then append the new data when it becomes available and call
//! // tokenize again.
//! ```
//!
//! # Encoding
//!
//! It is assumed that the data is UTF-8 encoded. There is no validation on the
//! actual data being tokenized (e.g. there could be control characters or other
//! data which may not be expected). Data validation is done by the library caller.
//!
//! # Well-formed vs. Malformed document processing
//!
//! The library should scan and evaluate well-formed XML documents correctly. For
//! XML documents which are not well-formed, the behavior is currently undefined. The library
//! does not error when scanning a malformed document.
//!
//! # Security Considerations
//!
//! The input is managed by the library user. If there is malformed input, the
//! tokenizing functions could never return a complete token.
//!
//! For instance, the input could start with a `<` but there is no closing `>`
//! character.
//!
//! In particular, if data is coming over the network and the data is being
//! stored in a buffer, the buffer may have unbounded growth if the buffer's
//! data is freed only if a complete token is found.
//!
//! [xml]: https://www.w3.org/TR/2006/REC-xml11-20060816/

#![cfg_attr(not(feature = "std"), no_std)]
#![cfg_attr(docsrs, feature(doc_cfg))]
#![warn(
    missing_copy_implementations,
    missing_debug_implementations,
    missing_docs,
    rust_2018_idioms,
    unused_lifetimes,
    unused_qualifications
)]

#[cfg(all(feature = "alloc", not(feature = "std")))]
extern crate alloc;

pub(crate) mod bytes;
mod lexer;
pub mod scanner;
pub mod token;

pub use lexer::{scan, Lexer};
